crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,265 @@
+
+const LOG_ENV_KEY: &str = "AIR_LOG";
+
+pub(crate) struct LogMessage {
+    contents: String,
+}
+
+pub(crate) type LogMessageSender = tokio::sync::mpsc::UnboundedSender<LogMessage>;
+pub(crate) type LogMessageReceiver = tokio::sync::mpsc::UnboundedReceiver<LogMessage>;
+
+pub(crate) struct LogState {
+    client: Client,
+    log_rx: LogMessageReceiver,
+}
+
+// Needed for spawning the loop
+unsafe impl Sync for LogState {}
+
+impl LogState {
+    pub(crate) fn new(client: Client) -> (Self, LogMessageSender) {
+        let (log_tx, log_rx) = unbounded_channel::<LogMessage>();
+        let state = Self { client, log_rx };
+        (state, log_tx)
+    }
+
+    /// Start the log loop
+    ///
+    /// Takes ownership of log state and start the low-latency log loop.
+    ///
+    /// We use `MessageType::LOG` to prevent the middleware from adding its own
+    /// timestamp and log level labels. We add that ourselves through tracing.
+    pub(crate) async fn start(mut self) {
+        while let Some(message) = self.log_rx.recv().await {
+            self.client
+                .log_message(MessageType::LOG, message.contents)
+                .await
+        }
+
+        // Channel has been closed.
+        // All senders have been dropped or `close()` was called.
+    }
+}
+
+// A log writer that uses LSPs logMessage method.
+struct LogWriter<'a> {
+    log_tx: &'a LogMessageSender,
+}
+
+impl<'a> LogWriter<'a> {
+    fn new(log_tx: &'a LogMessageSender) -> Self {
+        Self { log_tx }
+    }
+}
+
+impl Write for LogWriter<'_> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let contents = str::from_utf8(buf).map_err(|e| IoError::new(ErrorKind::InvalidData, e))?;
+        let contents = contents.to_string();
+
+        // Forward the log message to the latency sensitive log thread,
+        // which is in charge of forwarding to the client in an async manner.
+        self.log_tx
+            .send(LogMessage { contents })
+            .map_err(|e| IoError::new(ErrorKind::Other, e))?;
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+struct LogWriterMaker {
+    log_tx: LogMessageSender,
+}
+
+impl LogWriterMaker {
+    fn new(log_tx: LogMessageSender) -> Self {
+        Self { log_tx }
+    }
+}
+
+impl<'a> MakeWriter<'a> for LogWriterMaker {
+    type Writer = LogWriter<'a>;
+
+    fn make_writer(&'a self) -> Self::Writer {
+        LogWriter::new(&self.log_tx)
+    }
+}
+
+pub(crate) fn init_logging(
+    log_tx: LogMessageSender,
+    log_level: Option<LogLevel>,
+    client_info: Option<&ClientInfo>,
+) {
+    let log_level = resolve_log_level(log_level);
+
+    let writer = if client_info.is_some_and(|client_info| {
+        client_info.name.starts_with("Zed") || client_info.name.starts_with("Visual Studio Code")
+    }) {
+        BoxMakeWriter::new(LogWriterMaker::new(log_tx))
+    } else {
+        BoxMakeWriter::new(std::io::stderr)
+    };
```

Comment:
Doing what ruff does here. It seems like VS Code and Zed handle `window/logMessage` well, but perhaps not every lsp client does? So they specially approve ones that handle it well and send to stderr otherwise.